HiggsTools/Lilith interface

wkotlarski opened the pull request:
This PR contains a [HiggsTools](https://gitlab.com/higgsbounds/higgstools) and [Lilith](https://github.com/sabinekraml/Lilith-2) interfaces for FlexibleSUSY.

DONE:
- HiggsTools interface (with SLHA and WL outputs)
- Lilith interface (with SLHA output)

TODO:
- [ ] undetected BR is implemented. I defined as $\Gamma_{tot} - \Gamma_{inv} - SM$. check which SM channel have to be subtracted
- [ ] how to handle errors? A user request `HiggsTools` output (`FlexibleDecay` flag 6 set to 1). He/she then expects that the `HiggsTools` block will appear in the output.

wkotlarski commented:
- [ ] naming of files (consistent with `model_specific/SM`)
- [ ] naming of decay class (consistent with `model_specific/SM`)
- [ ] information about when was the code generated

wkotlarski commented:
@Expander, is there some smart way to construct a SM with Higgs of a given mass?

Expander commented:
> @Expander, is there some smart way to construct a SM with Higgs of a given mass?

I'm not sure. Maybe you could write
~~~.cpp
Standard_model sm(scale, loops, thresholds, g1, g2, g3, Lambdax, Yu, Yd, Ye, mu2, v);
~~~
where you adapt `Lambdax` such that the tree-level Higgs mass has the desired value?

wkotlarski commented:
Done with the SM Higgs mass. If you find some time I wouldn't mind if you checked my changes to the build system.

wkotlarski commented:
We should make sure that the build in SM is setup the same way as the BSM model. Currently, if you compute normalized effective couplings in the SM we don't get 1
```
coup  -11 (1.06028,0)
coup  -22 (1.00318,0)
coup  -33 (1.03276,0)
coup  -44 (1.00312,0)
coup  -55 (1.0118,0)
coup  -1111 (1.00299,0)
coup  -1313 (1.00299,0)
coup  -1515 (1.00299,0)
coup  -2424 (0.997968,0)
coup  2323 (0.993601,0)
coup  2322 (0.906263,0)
coup  2222 (0.654569,0)
coup  2121 (0.969308,0)
```

pathron commented:
> > @Expander, is there some smart way to construct a SM with Higgs of a given mass?
> 
> I'm not sure. Maybe you could write
> 
> ```c++
> Standard_model sm(scale, loops, thresholds, g1, g2, g3, Lambdax, Yu, Yd, Ye, mu2, v);
> ```
> 
> where you adapt `Lambdax` such that the tree-level Higgs mass has the desired value?

It would be nice if we could easily use the SM with Mh input and fixed in the BC, but I guess this would then miss the benefits of the hand written SM?

Expander commented:
> It would be nice if we could easily use the SM with Mh input and fixed in the BC, but I guess this would then miss the benefits of the hand written SM?

Good point. Would it help you to use the method `initialise_from_input(const softsusy::QedQcd&)` from the `Standard_model` class?

wkotlarski commented:
We already use `initialise_from_input(const softsusy::QedQcd&)` :sunglasses:
```c++
standard_model::Standard_model sm {};
sm.initialise_from_input(qedqcd);
auto match_Higgs_mass = [&sm, mass](Eigen::Matrix<double,1,1> const& x) {
   sm.set_Lambdax(x[0]);
   sm.solve_ewsb_tree_level();
   sm.calculate_DRbar_masses();
   sm.calculate_pole_masses();
   return std::abs(sm.get_physical().Mhh - mass);
};
```
I never know if the calculation of masses is ok as you have to call few commands in row in a proper order...

wkotlarski commented:
The difference in `h -> gamgam` between build in and generated SM comes from diagrams with W, charged Goldstones and ghosts. Could seems to be some difference in running W-masses between both SMs.

wkotlarski commented:
Found it :angry: Build in SM and autogenerated one have opposite signs in up-quark Yukawa definition (but not for down or lepton). After taking this into account we get the following effective couplings in the SM (so SM/SM)
```
coup  -11 (1.00319,0)
coup  -22 (1.0032,0)
coup  -33 (1.00316,0)
coup  -44 (1.00314,0)
coup  -55 (1.00312,0)
coup  -66 (0,0)
coup  -1111 (1.00301,0)
coup  -1313 (1.00301,0)
coup  -1515 (1.00301,0)
coup  -2424 (0.998382,0)
coup  2323 (0.993965,0)
coup  2322 (1.00561,0)
coup  2222 (1.01138,0)
coup  2121 (1.00714,0)
```

Expander commented on model_specific/SM/decays/standard_model_decay_table.cpp:
```diff
@@ -0,0 +1,81 @@
+}
+
+/*
+void SM_decay_table::print(std::ostream& ostr) const
```
Why is this function commented out? Some printing function would be very useful, I think, so we may keep it (in this or another form).

wkotlarski replied:
Uncommented in 6937ae2.

Expander replied:
Thanks for this!

Expander commented on src/decays/HiggsTools_interface.cpp:
```diff
@@ -0,0 +1,177 @@
+   std::ofstream hb_output("HiggsBounds.out");
+   hb_output << hbResult;
+   hb_output.close();
+   std::cout << "All applied limits: obsRatio (expRatio)\n";
```
In the end such `cout` statement should be removed or replaced by `VERBOSE_MSG` or similar, I think.

wkotlarski replied:
These particular ones will be completely removed when we decide on where to put the `HiggsBounds` output.

Expander replied:
> @wkotlarski : I am looking at the firsttodo item, regarding the naming of the files. Does this item refer to the pre-generated SM-specific files in `model_specific/SM/`?

Ok! I'm fine with the names of the new files.

Expander commented on src/decays/decay.cpp:
```diff
@@ -149,4 +149,155 @@ std::string strip_field_namespace(std::string const& s) {
    }
 }
 
+double hVV_4body(double *q2, size_t dim, void *params)
```
We may write `(double *q2, size_t /* dim */, ...)` to ignore the `dim` parameter.

wkotlarski replied:
Fixed in 790728a.

Expander commented on src/decays/decay_functions.cpp:
```diff
@@ -325,4 +325,27 @@ double sm_down_quark_masses(softsusy::QedQcd const& qedqcd, int n)
+
+double Delta_t(double z) noexcept {
+   const double lnz = std::log(z);
+   const double ln1mz = std::log(1.-z);
```
It would be better to write `std::log1p(-z)` here. (This also applies to the next two functions.)

Expander commented:
@wkotlarski : I am looking at the firsttodo item, regarding the naming of the files. Does this item refer to the pre-generated SM-specific files in `model_specific/SM/`?

Expander commented:
Regarding the HT output, I've tried to run the `SM` example of HT as so:
~~~
./examples/SM ../../hbdataset/ ../../hsdataset/
HiggsBounds result: allowed
	particle | obsRatio | expRatio | selected limit description
	---------|----------|----------|---------------------------
	    h    |   0.892  |   4.572  | LHC8 [vbfH,HW,Htt,H,HZ]>[bb,tautau,WW,ZZ,gamgam] from CMS-PAS-HIG-12-045 (CMS 17.3fb-1, M=(110, 600))

All applied limits: obsRatio (expRatio)
12045 LHC8 [vbfH,HW,Htt,H,HZ]>[bb,tautau,WW,ZZ,gamgam]: 0.892234 (4.57233)
13125353 LHC8 [vbfH,HW,Htt,H,HZ]>[ZZ]: 0.720629 (4.18302)
13003 LHC8 [H>WW]: 1.03812 (2.94213)
13121129 LHC8 [ggH,vbfH,HW,HZ]>[WW]: 0.969006 (2.90264)
20130131 LHC8 [vbfH,HW,Htt,H,HZ]>[ZZ]: 0.497169 (2.66269)
12077214 LHC8 [vbfH,HW,Htt,H,HZ]>[bb,tautau,WW,ZZ,gamgam]: 0.876553 (2.41663)
13001 LHC8 [vbfH,HW,Htt,H,HZ]>[gamgam]: 0.877541 (2.32813)
14122641 LHC8 [vbfH,HW,H,HZ]>[WW]: 0.745156 (2.14293)
14096212 LHC8 [HW,HZ]>[bb]: 0.773238 (1.23836)
13103687 LHC8 [HW,HZ]>[bb]: 0.495892 (0.985872)
13022 LHC8 [vbfH>WW]: 1.11008 (0.898529)
14029 2d likelihood {LHC8 [ggH>tautau], LHC8 [bbH>tautau]}: 0.572041 (0.486168)
180306553 2d likelihood {LHC13 [ggH>tautau], LHC13 [bbH>tautau]}: 0.39087 (0.43659)
150601010 LHC8 [vbfH>bb]: 0.173297 (0.370258)
13009 LHC8 [HW]>[tautau,WW]: 0.292585 (0.311466)
140960641 LHC8 [ggH>tautau]: 0.298379 (0.282466)
1708002122 LHC13 [ggH,vbfH,HW,HZ]>[Zgam]: 0.170955 (0.256433)
12051 LHC8 [HW,HZ]>[tautau]: 0.201234 (0.254861)
141066791 LHC8 [vbfH,HW,H,HZ]>[mumu]: 0.146973 (0.167406)
14067663 LHC8 [vbfH,HW,H,HZ]>[mumu]: 0.152313 (0.152109)
1810118221 LHC13 [ggH>bb]: 0.125878 (0.136561)
14023051 LHC8 [vbfH,HW,Htt,H,HZ]>[Zgam]: 0.0973134 (0.120504)
13075515 LHC8 [vbfH,HW,Htt,H,HZ]>[Zgam]: 0.119604 (0.113508)
1508014372 LHC8 [ggH>mumu]: 0.0479339 (0.0577272)
1911049682 LHC13 [Htt>mumu]: 0.0277653 (0.051305)
1911049681 LHC13 [Htt>mumu]: 0.0233079 (0.0455429)
1809059371 LHC13 [vbfH,HW,H,HZ]>[inv]: 0.00628192 (0.0079571)
190405105 LHC13 [HW,H,HZ]>[inv]: 0.00464969 (0.00711128)
180204329 LHC13 [HZ>cc]: 0.00972502 (0.0067327)
140960642 LHC8 [bbH>tautau]: 0.00504984 (0.00526057)
1809059372 LHC13 [vbfH,H]>[inv]: 0.00370107 (0.00491641)
510022 LEP [eeHZ]>[uu,dd,cc,ss,bb,gg]: 0.0039639 (0.00436639)
180906682 LHC13 [vbfH>inv]: 0.00248029 (0.00324655)
140413443 LHC8 [vbfH,HZ]>[inv]: 0.00177711 (0.00234341)
170809624 LHC13 [qqHZ>inv]: 0.00128611 (0.00221835)
140413441 LHC8 [vbfH>inv]: 0.00157687 (0.00209197)
14023244 LHC8 [HZ>inv]: 0.00147911 (0.00178355)
140413442 LHC8 [HZ>inv]: 0.00130151 (0.00127214)
150404324 LHC8 [HW,HZ]>[inv]: 0.0010549 (0.00103521)
1508014371 LHC8 [bbH>mumu]: 0.00086525 (0.000898036)
150201518 LHC8 [HW,H,HZ]>[inv]: 0.00076103 (0.000633503)
150608329 LHC8 [bbH>bb]: 0.00060761 (0.000632911)

 HiggsSignals chisq: 123.278 from 131 observables
~~~
If I see correctly, the output is a table of numbers in memory (not a file).

At first glance it seems to me that it would very nice if we could store the output in our SLHA  output, if there is a standard way (standard block) to do this. @wkotlarski and @pathron how do you usually store and use the HT output? Is there more information from HT that we can/should output? (I'm just wondering how the output is structured in order to decide whether it would be best to store it in an SLHA or a separate file.)

wkotlarski commented:
Yes, the data exist in memory. There are two main objects: `Higgs::Bounds` and `Higgs::Signals`.  `Higgs::Bounds` even has an implementation of an `<<` operator, so it can be used to print it to a terminal or a file directly. But those objects also contain a lot of different information which can be chosen to print or not.

wkotlarski commented on src/decays/decay_functions.cpp:
```diff
@@ -325,4 +325,27 @@ double sm_down_quark_masses(softsusy::QedQcd const& qedqcd, int n)
+double Delta_t(double z) noexcept {
+   const double lnz = std::log(z);
+   const double ln1mz = std::log(1.-z);
+   // eq. 2.5 of 2201.08139
```
Do you mean something like this?
```
/**
 * Calculates the Fermi constant \f$\sin\hat{\theta}_W\f$ as
 * defined in Eq. (C.3) from hep-ph/9606211 given theta,
 * the Z-boson pole mass and the DR-bar electromagnetic coupling
 *
 * @return value != 0 if an error has occured
 */
```

Expander replied:
Yes, something like this, at least with a short description of what is calculated and a reference with equation number.

wkotlarski commented:
Idea: if we want to have a block storing HT output, at least the `HiggsSignals` part is simple. We could write just one single number for $\chi^2$/n.o.d.f, putting extra details in the comment. Something like
```
1   1.30357   # 𝜒² = 170.768 from 131 observables
```

Expander commented:
> Idea: if we want to have a block storing HT output, at least the `HiggsSignals` part is simple. We could write just one single number for χ2/n.o.d.f, putting extra details in the comment. Something like
> 
> ```
> Block HiggsTools
> 1   1.30357   # HiggsSignals' 𝜒²/n.d.o.f, 𝜒² = 170.768 from 131 observables
> ```

That's a good idea. TBH, I'm not familiar with what would be useful to output. A new `HiggsTools` block is a good idea, but it could only store a set of scalar values (i.e. not vector or matrix like). At first glance it seems that the output from above seems to fit well into `HiggsTools` block. But I'm not sure if there is more (probably).

@pathron what output would be most useful for fitting programs like GAMBIT?

wkotlarski commented:
This is all that is needed from `HiggsSignals`
```
Block HiggsTools
1   1.30357   # HiggsSignals' 𝜒²/n.d.o.f, 𝜒² = 170.768 from 131 observables
```
`HiggsBounds` is a little more complicated. This is the minimal output
```
        particle | obsRatio | expRatio | selected limit description
        ---------|----------|----------|---------------------------
          Ah(1)  |  67.536  |  32.349  | 2d likelihood {LHC13 [ggH>tautau], LHC13 [bbH>tautau]} from 2002.12223 (ATLAS 139fb-1, M=(200, 2500))
          hh(1)  |  67.536  |  32.349  | 2d likelihood {LHC13 [ggH>tautau], LHC13 [bbH>tautau]} from 2002.12223 (ATLAS 139fb-1, M=(200, 2500))
```
If we want to make a block out of it, we'll have to chose which info to pick. Maybe something like this?
```
Block HiggsBounds
PDG   obsRatio   expRatio   # selected limit description
```

Expander commented:
> This is all that is needed from `HiggsSignals`
> 
> ```
> Block HiggsTools
> 1   1.30357   # HiggsSignals' 𝜒²/n.d.o.f, 𝜒² = 170.768 from 131 observables
> ```
> 
> `HiggsBounds` is a little more complicated. This is the minimal output
> 
> ```
>         particle | obsRatio | expRatio | selected limit description
>         ---------|----------|----------|---------------------------
>           Ah(1)  |  67.536  |  32.349  | 2d likelihood {LHC13 [ggH>tautau], LHC13 [bbH>tautau]} from 2002.12223 (ATLAS 139fb-1, M=(200, 2500))
>           hh(1)  |  67.536  |  32.349  | 2d likelihood {LHC13 [ggH>tautau], LHC13 [bbH>tautau]} from 2002.12223 (ATLAS 139fb-1, M=(200, 2500))
> ```
> 
> If we want to make a block out of it, we'll have to chose which info to pick. Maybe something like this?
> 
> ```
> Block HiggsBounds
> PDG   obsRatio   expRatio   # selected limit description
> ```

Ah, I see! Your proposal for a `HiggsBounds` block seems optimal for this data.

pathron commented:
> 

Whenever I have used HiggsBounds / Higgs Signals I am just interfacing directly to the functions outputting the likelihood as far as I recall.  Gambit also avoids direct use of SLHA files so I am not sure I have a useful answer here.  Regarding content, while the chi^2 per degree of freedom is most useful I think outputting the actual chi-squared and the associated number of degrees of freedom as entries would be preferable.  I guess they could be scraped from the comments but its better to have them as entries IMO.

pathron commented:
Sorry stupid question, but don't HiggsBounds and HiggsSignals themselves accept SLHA input and then provide an SLHA output with some new output blocks for their results?  And presumably HiggsTools as well. If so shouldn't we just be taking over their blocks verbatim?

wkotlarski commented:
> Sorry stupid question, but don't HiggsBounds and HiggsSignals themselves accept SLHA input and then provide an SLHA output with some new output blocks for their results? And presumably HiggsTools as well. If so shouldn't we just be taking over their blocks verbatim?

The only thing that the new manual says is 

> Moreover, SLHA files or HiggsBounds data files [6] can be used as input via the Python interface.

But in general I think that the answer is no. We want to run in effective coupling mode (parametrize Higgs sector via eff. couplings, not brs and xsections). This information is not part of the standard.

wkotlarski commented:
I've implemented the output as follows
```
Block HIGGSSIGNALS
     1     1.31000000E+02   # number of degrees of freedom
     2     1.70767811E+02   # 𝜒²
```

wkotlarski commented:
@Expander, @Peter: as stated [here](https://gitlab.com/higgsbounds/higgstools#obtaining-the-dataset) `HiggsSignals` and `HiggsBounds` databases are located outside of `HiggsTools` instalation. How should we pass it's location to `FlexibleSUSY`. It doesn't have to known at compile time, but on the other hand specyfing it at runtime might be a pain. One solution could be to add `--with-higgssignals-dataset` and `--with-higgsbounds-dataset` options to the `configure` script.

Expander commented:
> @Expander, @peter: as stated [here](https://gitlab.com/higgsbounds/higgstools#obtaining-the-dataset) `HiggsSignals` and `HiggsBounds` databases are located outside of `HiggsTools` instalation. How should we pass it's location to `FlexibleSUSY`. It doesn't have to known at compile time, but on the other hand specyfing it at runtime might be a pain. One solution could be to add `--with-higgssignals-dataset` and `--with-higgsbounds-dataset` options to the `configure` script.

I'd vote for passing it via a command line option at run-time, maybe like so:
~~~.sh
./run_MSSM.x --with-higgssignals-dataset=<path> --with-higgsbounds-dataset=<path>
~~~
I think passing it to `configure` would be less convenient, because then the code would be needed to be re-compiled, which I think would be unnecessary if one would just like to use another data set.

pathron commented:
How does one then specify it if you are not using the command line interface?  For example if you are scanning with multinest.

wkotlarski commented:
> I think passing it to `configure` would be less convenient, because then the code would be needed to be re-compiled, which I think would be unnecessary if one would just like to use another data set.

I think that this is not a practical issue. Those databases actually change less than `HiggsTools` itself. Last change was made 5 months ego. There are also no alternative databases, so that you might want to run your code with different ones to compare. So at least for now I'd say that specifying them at runtime is an overkill.

Expander commented:
> I think that this is not a practical issue. Those databases actually change less than `HiggsTools` itself. Last change was made 5 months ego. There are also no alternative databases, so that you might want to run your code with different ones to compare. So at least for now I'd say that specifying them at runtime is an overkill.

That's of course a valid point. However I had also the following scenarios in mind:
1. The user is moving the executable to another machine, where the databases are at a different  place
2. The user moves the databases to another directory

I think in particular 1. may happen. So, maybe we should at least provide both options: Setting the default directories via `configure` and also allow the user to override them when running the executables. What do you think?

wkotlarski commented:
I don't want to do both of those things so I've implemented the command line option (d433aa2). We need to catch the `SetupError` in `run.cpp` and decide where to put it.

Expander commented:
Thanks for implementing this!

Do these tests fail, because we currently throw an exception if the HT databases are not found? This is indeed not good and should be improved. One option may be to simply not calculate the HT observables if the database is not found. In this case we may output `0` or `NaN` or so. What do you think?

Expander commented on src/decays/HiggsTools_interface.cpp:
```diff
@@ -0,0 +1,189 @@
+
+
+   // HiggsBounds
+   struct stat buffer;
```
`sys/stat.h` is UNIX-specific. I think an OS-independent solution would be better, like so:
~~~.cpp
#include <boost/filesystem.hpp>

if (!boost::filesystem::exists("myfile.txt")) {
   ...
}
~~~

wkotlarski replied:
Now that I think of this, since `HiggsTools` interface requires C++17, we can use a C++17 `filesystem` module.

wkotlarski commented:
Hi @Expander. Unfortunately, I need some input from you. What is the correct way to create a built in standard model in a way that would be equivalent to the auto generated one? In short, I need a two scale version of the built in sm. I see that something similar is done in `test/test_SM_low_scale_constraint.cpp`, but there we set parameters by hand using the `setup_SM_const` function. Maybe you'd have time for a short skype?

wkotlarski commented:
Friday is great. Just not between 11:15 and ~12:30. Other than that I'm free. Thanks Alex. I appreciate the help.

Expander commented:
I probably can't make it before 6pm. Would that be ok with you?

wkotlarski commented:
Sure.

wkotlarski commented:
@Expander, we are quite close to making this working too. The normalized (to SM) effective couplings are now
```
Block NORMALIZEDEFFHIGGSCOUPLINGS
       25           0           0     4.02013692E-03   #
       25          -1           1     1.00000006E+00   #
       25          -2           2     1.00000006E+00   #
       25          -3           3     1.00000013E+00   #
       25          -4           4     9.99997015E-01   #
       25          -5           5     9.99904980E-01   #
       25          -6           6     0.00000000E+00   #
       25         -11          11     1.00000006E+00   #
       25         -13          13     1.00000006E+00   #
       25         -15          15     1.00000004E+00   #
       25         -24          24     9.99973043E-01   #
       25          23          23     9.99983629E-01   #
       25          21          21     1.00000010E+00   #
       25          22          22     1.00006125E+00   #
       25          23          22     9.98553609E-01   #
```
The difference for `Z\gamma` comes from the difference in computed $Z$ pole mass. In `ZZ` I've the mass from `QedQCD`. We should decide whether this is fine. It will break if Z-multiplet has also other particles. But I cannot find any such model in SARAH. In $Z'$ models the $Z'$ always lives separately. If we go with not taking Z mass from `QedQCD` then we should thing about this mass difference.

The crucial part to check is this initialization:
```cpp
standard_model::Standard_model sm {};    
sm.initialise_from_input(qedqcd);    
sm.set_physical_input(physical_input);    
sm.set_pole_mass_loop_order(static_cast<int>(spectrum_generator_settings.get(Spectrum_generator_settings::pole_mass_loop_order)));    
sm.set_ewsb_loop_order(static_cast<int>(spectrum_generator_settings.get(Spectrum_generator_settings::ewsb_loop_order)));    
sm.set_precision(spectrum_generator_settings.get(Spectrum_generator_settings::precision));    
sm.set_threshold_corrections(spectrum_generator_settings.get_threshold_corrections());    
sm.set_loop_corrections(spectrum_generator_settings.get_loop_corrections());    
sm.set_loops(static_cast<int>(spectrum_generator_settings.get(Spectrum_generator_settings::beta_loop_order)));    
Loop_corrections loop_corrections_;    
loop_corrections_.higgs_at_as = spectrum_generator_settings.get(Spectrum_generator_settings::higgs_2loop_correction_at_as);    
loop_corrections_.higgs_ab_as = spectrum_generator_settings.get(Spectrum_generator_settings::higgs_2loop_correction_ab_as);    
loop_corrections_.higgs_at_at = spectrum_generator_settings.get(Spectrum_generator_settings::higgs_2loop_correction_at_at);    
loop_corrections_.higgs_atau_atau = spectrum_generator_settings.get(Spectrum_generator_settings::higgs_2loop_correction_atau_atau);    
loop_corrections_.top_qcd = spectrum_generator_settings.get(Spectrum_generator_settings::top_pole_qcd_corrections);    
loop_corrections_.higgs_at_as_as = spectrum_generator_settings.get(Spectrum_generator_settings::higgs_3loop_correction_at_as2);    
loop_corrections_.higgs_ab_as_as = spectrum_generator_settings.get(Spectrum_generator_settings::higgs_3loop_correction_ab_as2);    
loop_corrections_.higgs_at_at_as = spectrum_generator_settings.get(Spectrum_generator_settings::higgs_3loop_correction_at2_as);    
loop_corrections_.higgs_at_at_at = spectrum_generator_settings.get(Spectrum_generator_settings::higgs_3loop_correction_at3);    
loop_corrections_.higgs_at_as_as_as = spectrum_generator_settings.get(Spectrum_generator_settings::higgs_4loop_correction_at_as3);    
sm.set_loop_corrections(loop_corrections_);    
    
// set SM λ such that mhSM == mass    
auto match_Higgs_mass = [&sm, mass](double x) {    
   sm.set_Lambdax(x);    
   sm.calculate_DRbar_masses(); // internaly calls solve_ewsb_tree_level()    
   sm.solve_ewsb();    
   sm.calculate_Mhh_pole();    
   return std::abs(sm.get_physical().Mhh - mass);    
};
```
which is followed by minimization of `match_Higgs_mass`. I have two questions

1. I try to match the `Spectrum_generator_settings` between models. Does this looks fine to you?
2. I'm calling `initialise_from_input` (which internally sets `Lambdax`) and then override `Lambdax`. Is it ok?

Expander commented:
Sorry, I'm not quite sure I understand.

If we have genuine Z-Z' mixing, are the ratios that we compute well-defined? If so, then there should be a way to chose which of the Z-like bosons is the SM one to be used in the calculation, right? To provide such an option, we could do something similar to what we do with the SM-like Higgs.

On the other side: In current FS we only support models that contain the SM as a subset (including the gauge group). So, genuine Z-Z' mixing with some non-SM gauge group is currently not supported by FS.

To clarify this we could look into the E6SSM, which contains a Z' that mixes with the others. How would you proceed in this model?

wkotlarski commented:
Maybe I'm missing something, but if I look at [E6SSM](https://github.com/FlexibleSUSY/FlexibleSUSY/blob/master/sarah/E6SSM/particles.m) then I see two EWSB states: `VZ` and `VZp`. Doesn't it mean that they are **not** part of the same multiplet?

Expander commented:
You are right to some extend: The final masses are "pulled apart". Actually we have:
~~~.cpp
void CLASSNAME::calculate_MVPVZVZp()
{
   const auto mass_matrix_VPVZVZp(get_mass_matrix_VPVZVZp());
   Eigen::Array<double,3,1> MVPVZVZp;

   fs_diagonalize_hermitian(mass_matrix_VPVZVZp, MVPVZVZp, ZZ);

   ZZ.transposeInPlace();
   normalize_to_interval(ZZ);

   MVPVZVZp = AbsSqrt(MVPVZVZp);

   MVP = 0.;
   MVZ = MVPVZVZp(1);
   MVZp = MVPVZVZp(2);
}
~~~

wkotlarski commented:
Great. So the question is whether it's ok to do it like this:

1. setup `Standard_model` with `Spectrum_generator_settings` copied from the BSM model
2. initialize `Standard_model` from `QedQCD`
3. set $\lambda$ giving the proper Higgs pole mass matching the BSM Higgs pole mass. This means we compute a mass $m_h$ in range 1-650 GeV at the renormalization scale $\mu = m_Z$.
4. Finally, for decays we run running parameters to the $\mu = m_h$ (again, $1 \lesssim m_h \lesssim 650$ GeV)

Expander commented:
Yes, this sounds as a very reasonable approach! Step 3 means that you're somehow performing a matching to the BSM model.

wkotlarski commented:
The discrepancy will grow with the Higgs mass.

For SM with $m_h = 124.8$ GeV (which happens for $\lambda =  0.192$) couplings normalized to SM are
```
       25          -1           1     1.00000006E+00   #
       25          -2           2     1.00000006E+00   #
       25          -3           3     1.00000013E+00   #
       25          -4           4     9.99997015E-01   #
       25          -5           5     9.99904980E-01   #
       25          -6           6     0.00000000E+00   #
       25         -11          11     1.00000006E+00   #
       25         -13          13     1.00000006E+00   #
       25         -15          15     1.00000004E+00   #
       25         -24          24     9.99973043E-01   #
       25          23          23     9.99983629E-01   #
       25          21          21     1.00000010E+00   #
       25          22          22     1.00006125E+00   #
       25          23          22     1.00002562E+00   #
```
Meanwhile, for SM with $m_h = 393$ GeV ($\lambda = 3.5$) we get
```
       25          -1           1     9.99473400E-01   #
       25          -2           2     9.99473400E-01   #
       25          -3           3     9.99473410E-01   #
       25          -4           4     9.99473101E-01   #
       25          -5           5     9.99461388E-01   #
       25          -6           6     1.02237707E+00   #
       25         -11          11     9.99473416E-01   #
       25         -13          13     9.99473416E-01   #
       25         -15          15     9.99473491E-01   #
       25         -24          24     1.00052801E+00   #
       25          23          23     1.00052782E+00   #
       25          21          21     9.99590846E-01   #
       25          22          22     1.01083809E+00   #
       25          23          22     1.00182622E+00   #
```
This isn't a huge practical problem. This influences constraints of heavy (BSM) Higgses. There the precision of 2% is probably acceptable. For the SM Higgs we're getting better than 0.1% accuracy.

Expander commented:
Is this below the uncertainty that you've described, as these ratios are an approximation anyway?

wkotlarski commented:
Hi @Expander. I want to finish this PR asap. The only thing that was left was the mathematica interface. There's a question about mathematica interface to `HiggsTools` call, though. We pass the location of `HiggsSignals` and `HiggsBounds` database at runtime. The idea is to have a call like
```mathematica
FSSMCallHiggsTools[handle, "HBDataSet", "HSDataSet"];
```
As far as I understand, for this I cannot use `MLINK`. So I went with this
```cpp
  DLLEXPORT int FS@ModelName@CallHiggsTools(
     WolframLibraryData libData, mint Argc, MArgument *Args, MArgument Res)
```
This works as far as pasing args goes, but I don't know how to return a complicated output from `FS@ModelName@CallHiggsTools`, in a way similar how I'd do it using `MLINK`. Do you know how this could be achieved?

Expander commented:
At first glance it seems to me that getting a string from an `MLINK` should be possible. I've looked into the C header `mathlink.h` (which is installed in `SystemFiles/Links/MathLink/DeveloperKit/Linux-x86-64/CompilerAdditions` on my machine). There I found
~~~.h
MLDECL( int,   MLGetString,           ( MLINK mlp, const char **sp));
~~~
So, it seems to me that one could write a function like the following one (not tested):
~~~.cpp
std::string get_string_from(MLINK link)
{
   const char* c_string = 0;

   if !(MLGetString(link, &c_string)) {
      std::cerr << "Error: argument is not a string." << std::endl;
   }
   
   return c_string;
}
~~~
Would that help?

If you want to go with your original function signature, where you take `MArgument Res` as last argument, you'd need to write the output to `Res`, I think. I found this demo script in `LibraryLink/LibraryResources/Source/demo_string.c` (there are more examples there):
~~~.c
DLLEXPORT int encodeString(WolframLibraryData libData, mint Argc, MArgument *Args, MArgument Res) 
{
        mint i = 0, shift;
        
        if (string) 
                libData->UTF8String_disown(string);

        string = MArgument_getUTF8String(Args[0]);
        shift = MArgument_getInteger(Args[1]);

        /* Find shift mod 127 so we only 
           deal with positive numbers below */
        shift = shift % 127;
        if (shift < 0) 
                shift += 127;

        shift -= 1; 
                
        while (string[i]) {
                mint c = (mint) string[i];
                /* Error for non ASCII string */
                if (c & 128) return LIBRARY_FUNCTION_ERROR;
                c = ((c + shift) % 127) + 1;
                string[i++] = (char) c;
        }
        MArgument_setUTF8String(Res, string);
        return 0;
}
~~~
It seems to me that the line `MArgument_setUTF8String(Res, string);` moves the modified string to the output (although I'm not 100% sure).

wkotlarski commented:
Thanks. Your comment was very useful. The only problem I'm still left with is that it's not possible to create a `List` with a string
```mathematica
MLPutFunction(link, "List", 1);
MLPut(link, " ");
```
`MLPut` cannot put an empty string or a string with a space in it. 

1. "" doesn't work
2. "a" does work
3. "a " doesn't

etc.

Expander commented:
Could you please try MLPutString? 

https://reference.wolfram.com/language/ref/c/MLPutString.html

wkotlarski commented:
In the meantime I've slightly tweaked the output and would now like to use a rule like so
```mathematica
MLPutRuleTo(link, "a ", "bestAnalysis");
```
and this fails.

Expander commented:
You are right, a rule is much better.

If you want to create a rule of the form `String -> String`, I think you could achieve that as follows:
~~~.cpp
MLPutFunction(link, "Rule", 2);
MLPutString(link, "a ");
MLPutString(link, "bestAnalysis");
~~~
Returning a string (instead of a symbol) is a good choice for file names, for example. If the returned object should be used in a symbolic manipulation, a symbol would be better, I think.

wkotlarski commented:
Thanks for all the help! I think we're mostly done. There's one broken test that I'll try to fix today. Would you be able to find some time to go through this PR? I'd like to merge it in a not-to-distant future.

Expander commented on configure:
```diff
@@ -173,6 +173,8 @@ boost_lib_dir=""
 eigen_inc_dir=""
 BOOSTFLAGS=""
+HIGGSTOOLSFLAGS="-Imodel_specific/SM"
+HIGGSTOOLSLIBS="model_specific/SM/libmodel_specific_SM.a"
```
Please use `${STATIC_LIB_EXT}` instead of `.a`, i.e.
~~~.sh
HIGGSTOOLSLIBS="model_specific/SM/libmodel_specific_SM.${STATIC_LIB_EXT}"
~~~
(This may also apply to your other changes below.)

wkotlarski replied:
Done in de968f92a015185a0b76f2e9d39007b23685a3d5.

Expander commented on configure:
```diff
@@ -2214,10 +2216,10 @@ check_higgstools_libs() {
             if [ -z "$higgstools_lib_dir" ]; then
                 # assume that the linker will look in the default paths
-                HIGGSTOOLSLIBS="-Wl,-Bdynamic -$HIGGSTOOLSLIBS_"
+                HIGGSTOOLSLIBS="model_specific/SM/libmodel_specific_SM.a -Wl,-Bdynamic -$HIGGSTOOLSLIBS_"
```
I think there is a mistake here: We're in the branch where the library extension is `.so` or `.dylib`. So we should probably have
~~~
HIGGSTOOLSLIBS="model_specific/SM/libmodel_specific_SM.${extension} -Wl,-Bdynamic -$HIGGSTOOLSLIBS_"
~~~

wkotlarski replied:
This branching is not about `model_specific_SM`. It's about `HiggsTools` and whether `HiggsTools` is compiled as shared or static library. The build in SM is always compiled as static library, right?

Expander replied:
Sorry, I misunderstood what the branch is about.

> The build in SM is always compiled as static library, right?

No. Actually whether shared or static libraries are created is controlled by `./configure --enable_shared_libs`. In `model_specific/SM/module.mk` we have
~~~.mk
LIB_model_specific_SM := $(DIR)/lib$(MODNAME)$(MODULE_LIBEXT)
~~~
where `MODULE_LIBEXT` is defined in `configure` to be the file extension of a static or shared library. (This reminds me that we should test that FS builds with static and shared libraries before we release.)

wkotlarski replied:
Tbh, I have no idea what I'm doing with this. There's `--enable-static` and `--enable-shared-libs`. My understanding is that the first controls how external libs are linked, the second which libs do we create

`--enable-static` is broken and that was long before  I got to it
```
Building models/SM/run_cmd_line_SM.x
/usr/lib64/gcc/x86_64-suse-linux/14/../../../../x86_64-suse-linux/bin/ld: cannot find -lcollier: No such file or directory
```

This is what I came up with for a moment
```bash
if [ "x$enable_static" = xyes ]; then
    if [ -z "$higgstools_lib_dir" ]; then
        HIGGSTOOLSLIBS="-Wl,-Bstatic -$HIGGSTOOLSLIBS_"
    else
        HIGGSTOOLSLIBS="-L$higgstools_lib_dir -Wl,-rpath,$higgstools_lib_dir,-Bstatic $HIGGSTOOLSLIBS_"
    fi
else
   HIGGSTOOLSLIBS="-L$higgstools_lib_dir ${HIGGSTOOLSLIBS_}"
fi
HIGGSTOOLSLIBS="${HIGGSTOOLSLIBS} model_specific/SM/libmodel_specific_SM${MODULE_LIBEXT}"
```
This does link `model_specific_SM` correctly (both with and w/o `--enable-shared-libs`) . It also links `HiggsTools` in shared mode correctly. The problem is with this internal block which triggers in static mode
```bash
if [ -z "$higgstools_lib_dir" ]; then
   HIGGSTOOLSLIBS="-Wl,-Bstatic -$HIGGSTOOLSLIBS_"
else
   HIGGSTOOLSLIBS="-L$higgstools_lib_dir -Wl,-rpath,$higgstools_lib_dir,-Bstatic $HIGGSTOOLSLIBS_"
fi
```

Expander commented on configure:
```diff
@@ -2202,10 +2204,10 @@ check_higgstools_libs() {
             if [ -z "$higgstools_lib_dir" ]; then
                 # assume that the linker will look in the default paths
-                HIGGSTOOLSLIBS="-Wl,-Bstatic -$HIGGSTOOLSLIBS_ -Wl,-Bdynamic"
+                HIGGSTOOLSLIBS="-Wl,-Bstatic -$HIGGSTOOLSLIBS_ model_specific/SM/libmodel_specific_SM.a -Wl,-Bdynamic"
```
I think it would be better to write
~~~
HIGGSTOOLSLIBS="-Wl,-Bstatic -$HIGGSTOOLSLIBS_ model_specific/SM/libmodel_specific_SM.${extension} -Wl,-Bdynamic"
~~~
instead of the hard-coded `.a` extension

wkotlarski replied:
Done in 97692f0677cd295dffcd1386dd95f9bff59b1388.

Expander commented on configure:
```diff
@@ -2214,10 +2216,10 @@ check_higgstools_libs() {
             else
                 # assume that the linker will look in the default paths
-                HIGGSTOOLSLIBS="-L$higgstools_lib_dir -Wl,-rpath,$higgstools_lib_dir,-Bdynamic $HIGGSTOOLSLIBS_"
+                HIGGSTOOLSLIBS="model_specific/SM/libmodel_specific_SM.a -L$higgstools_lib_dir -Wl,-rpath,$higgstools_lib_dir,-Bdynamic $HIGGSTOOLSLIBS_"
```
Same here: We should have 
~~~
HIGGSTOOLSLIBS="model_specific/SM/libmodel_specific_SM.${extension} -L$higgstools_lib_dir -Wl,-rpath,$higgstools_lib_dir,-Bdynamic $HIGGSTOOLSLIBS_"
~~~
if I understand correctly.

Expander commented on meta/Decays.m:
```diff
@@ -928,6 +928,14 @@ would violate Ward identity (as at the level of dim 4
+                          DeleteCases[{TreeMasses`GetHiggsBoson[], TreeMasses`GetPseudoscalarHiggsBoson[], TreeMasses`GetChargedHiggsBoson[] /. Susyno`LieGroups`conj->Identity}, Null],
+                          particle /. Susyno`LieGroups`conj->Identity
+                       ],
+                       "dm->set_use_pole_higgs_mixings(static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::use_pole_higgs_mixings)));\n",
```
Personally I'd like to write `x != 0.0` instead of `static_cast<bool>(x)`, because I find it more readable.

wkotlarski replied:
I could omit the cast entirely and just rely on auto conversion. Would that be ok?

wkotlarski replied:
Done in 58648c9e1e55c629742b75fd3a197eb4debc9c15.

Expander commented on meta/Decays.m:
```diff
@@ -1104,18 +1129,28 @@ would violate Ward identity (as at the level of dim 4
+                            "0.",
+                            If[IsZBoson[First[GetFinalState[decay]]],
+                               "qedqcd.displayPoleMZ()",
+                               paramsStruct <> ".physical_mass<" <> CXXDiagrams`CXXNameOfField[First[GetFinalState[decay]]] <> ">(idx_2)"
```
I'm wondering if it is guaranteed that the variable `idx_2` is always defined. (Maybe it is.)

wkotlarski replied:
Yes. There are always exactly 3 arguments called `idx_1`, `idx_2` and `idx_3`.

Expander replied:
Ah, ok!

Expander commented on meta/FSMathLink.m:
```diff
@@ -303,6 +348,38 @@
+                  "MLPutFunction(link, \"List\", 2);\n" <>
+                  "MLPut(link, " <> ToString[#[[1]]] <> ");\n" <>
+                  "MLPut(link, " <> ToString[#[[2]]] <> ");\n" <>
+                  "MLPut(link, d." <> #[[3]] <> ".second);\n")& /@ {{-1, 1, "dd"}, {-2, 2, "uu"}, {-3, 3, "ss"}, {-4, 4, "cc"}, {-5, 5, "bb"}, {-6, 6, "tt"}, {-11, 11, "ee"}, {-13, 13, "mumu"}, {-15, 15, "tautau"}, {-24, 24, "WW"}, {23, 23, "ZZ"}, {22, 23, "Zgam"}, {21, 21, "gg"}, {21, 22, "gamgam"}}
```
Would you mind moving this longish list of particles to a separate (local) variable? Then this line would be shorter and the number `14` appearing above could be replaced by `ToString[Length[particleList]]`.

wkotlarski replied:
Done in 93c434cb08416484495b9739bc749fad979aa274.

wkotlarski replied:
Fixed in 93c434cb08416484495b9739bc749fad979aa274.

Expander replied:
Thanks for this!

Expander commented on src/decays/specializations/Ah/decay_Ah_to_gg.inc:
```diff
@@ -18,9 +18,17 @@ double CLASSNAME::get_partial_width<PseudoscalarHiggs, Gluon, Gluon>(
+   if (Nf < 5) {
+      problems.add_warning(
+         create_process_string<PseudoscalarHiggs,Gluon,Gluon>(in_idx, out1_idx, out2_idx)
+         + ": Cannot determine the number of active quark flavours. Disabling higher-order corrections."
```
Is this error message correct? From looking at the if-statement it seems that the error is that `mAh` is actually too small.

Expander commented on src/decays/specializations/H/decay_H_to_ubaru.inc:
```diff
@@ -177,7 +177,24 @@ double CLASSNAME::get_partial_width<Higgs, bar<UpTypeQuark>::type, UpTypeQuark>(
+            boost::hana::unpack(bar<UpTypeQuark>::pdgids, _to_array<bar<UpTypeQuark>::numberOfGenerations>).at(indexOut1.at(0)),
+            boost::hana::unpack(UpTypeQuark::pdgids, _to_array<UpTypeQuark::numberOfGenerations>).at(indexOut2.at(0))
+         },
+         std::pair<std::string const&, std::complex<double>> {
```
The first argument of this `pair` is a reference to a string. Do you know whether the string lives long enough when passing the pair to `add_coupling()`?

wkotlarski replied:
You're right. I wasn't thinking. Done in 7a425d43beebf4ac277b67e30046be3c7fea0909.

Expander replied:
Thanks for this! Now that it is changed, I'm wondering if `add_coupling` could take the `std::pair` as a `const` references. That could prevent a string copy.

wkotlarski replied:
Done in 377f66666f5cd3530247cee671575d772a8291fe.

Expander commented:
Wow, this seems to have been a lot of work, thanks for implementing this!

A few questions, though:
1. It seems we don't have any tests for the HiggsTools and Lilith output. Naively I'd have thought that a simple test for some model where the results are known would be good. I.e. a simple test that ensures that the output is reasonable.
2. Would you mind testing (maybe just once on your machine) that FS builds with dynamic libraries as well (`--enable-shared-libs`)?

wkotlarski commented:
I've fixed the error with `--enable-shared-libs`. I had to add new libraries to `test/SOFTSUSY/module.mk`. And that was it. I will ofc add some tests for HiggsTools/Lilith interface.

wkotlarski commented:
I've fixed all the original tests that I've broken. Regarding new tests, I don't think that testing HiggsTools and Lilith is a good idea. We would be testing an external code. So anytime HiggsTools and Lilith is updated, those tests would have to change. Instead, I propose to test the input to HiggsTools and Lilith that we produce, namely the effective couplings normalized to SM values. I propose 2 tests:

1. Test effc in the SM. We expected to get as close to 1 as possible.
2. Make one test in the BSM model (like the MRSSM)

In the SM this looks like
```
   BOOST_CHECK_CLOSE(effc[0].gg.second, 1, 0.08);     // 0.08%    
   BOOST_CHECK_CLOSE(effc[0].gamgam.second, 1, 0.4);  // 0.4%    
   BOOST_CHECK_CLOSE(effc[0].Zgam.second, 1, 0.3);    // 0.3%    
    
   BOOST_CHECK_CLOSE(effc[0].ZZ.second, 1, 0.07);     // 0.07%    
   BOOST_CHECK_CLOSE(effc[0].WW.second, 1, 0);        // 8%    
    
   BOOST_CHECK_CLOSE(std::real(effc[0].ee.second),     1, 0.04);  // 0.04%    
   BOOST_CHECK_CLOSE(std::real(effc[0].mumu.second),   1, 0.04);  // 0.04%    
   BOOST_CHECK_CLOSE(std::real(effc[0].tautau.second), 1, 0.04);  // 0.04%    
    
   BOOST_CHECK_CLOSE(std::real(effc[0].bb.second), 1, 0.03);  // 0.03%    
   BOOST_CHECK_CLOSE(std::real(effc[0].cc.second), 1, 0.03);  // 0.03%    
   BOOST_CHECK_CLOSE(std::real(effc[0].ss.second), 1, 0.04);  // 0.04%    
   BOOST_CHECK_CLOSE(std::real(effc[0].dd.second), 1, 2);     // 2%    
   BOOST_CHECK_CLOSE(std::real(effc[0].uu.second), 1, 6);     // 6%    
```

I used `BOOST_CHECK_CLOSE` so the precision means an absolute difference in %. You can see that most effc are within 1 at around 0.1% level or better. The problem is with the $W^+W^-$ channel where the difference is 7-8%. Some time ago we changed the built in SM so that the $W$-mass returned from the `calculate_pole_masses` function is the one from the fit. But in the autogenerated one we do a pure one-loop calculation. As far as I remember, the autogenerated models store their "correct" $m_W$ in the `QedQCD` class. Do you think it would be enough to use this mass instead of the one from `calculate_pole_masses` to solve this problem.

Expander commented:
> Regarding new tests, I don't think that testing HiggsTools and Lilith is a good idea. We would be testing an external code. So anytime HiggsTools and Lilith is updated, those tests would have to change.

You are right, that's a good point. My thought was to make sure, that we have implemented the interface to HiggsTools and Lilith correctly (i.e. passing the correct couplings to the correct blocks etc.). But maybe it is sufficient to check this only once, for example the paper.

> Instead, I propose to test the input to HiggsTools and Lilith that we produce, namely the effective couplings normalized to SM values. I propose 2 tests:
> 
>     1. Test effc in the SM. We expected to get as close to 1 as possible.
> 
>     2. Make one test in the BSM model (like the MRSSM)
> 
> In the SM this looks like
> 
> ```
>    BOOST_CHECK_CLOSE(effc[0].gg.second, 1, 0.08);     // 0.08%    
>    BOOST_CHECK_CLOSE(effc[0].gamgam.second, 1, 0.4);  // 0.4%    
>    BOOST_CHECK_CLOSE(effc[0].Zgam.second, 1, 0.3);    // 0.3%    
>     
>    BOOST_CHECK_CLOSE(effc[0].ZZ.second, 1, 0.07);     // 0.07%    
>    BOOST_CHECK_CLOSE(effc[0].WW.second, 1, 0);        // 8%    
>     
>    BOOST_CHECK_CLOSE(std::real(effc[0].ee.second),     1, 0.04);  // 0.04%    
>    BOOST_CHECK_CLOSE(std::real(effc[0].mumu.second),   1, 0.04);  // 0.04%    
>    BOOST_CHECK_CLOSE(std::real(effc[0].tautau.second), 1, 0.04);  // 0.04%    
>     
>    BOOST_CHECK_CLOSE(std::real(effc[0].bb.second), 1, 0.03);  // 0.03%    
>    BOOST_CHECK_CLOSE(std::real(effc[0].cc.second), 1, 0.03);  // 0.03%    
>    BOOST_CHECK_CLOSE(std::real(effc[0].ss.second), 1, 0.04);  // 0.04%    
>    BOOST_CHECK_CLOSE(std::real(effc[0].dd.second), 1, 2);     // 2%    
>    BOOST_CHECK_CLOSE(std::real(effc[0].uu.second), 1, 6);     // 6%    
> ```
> 
> I used `BOOST_CHECK_CLOSE` so the precision means an absolute difference in %. You can see that most effc are within 1 at around 0.1% level or better.

Excellent idea, please go for it!

> The problem is with the W + W − channel where the difference is 7-8%. Some time ago we changed the built in SM so that the W -mass returned from the `calculate_pole_masses` function is the one from the fit. But in the autogenerated one we do a pure one-loop calculation. As far as I remember, the autogenerated models store their "correct" m W in the `QedQCD` class. Do you think it would be enough to use this mass instead of the one from `calculate_pole_masses` to solve this problem.

Hmm, at first sight I'm not so happy about using the W mass from `QedQcd`. Storing the W mass in `QedQcd` in the first place is already a hack. Then we'd rely on this hack, which is not so nice. Furthermore, `QedQcd` is only filled after the iteration between the scales is finished. If one would use the `mass_eigenstates` class w/o an iteration before, then the value of the W mass stored in `QedQcd` would not be properly set.

Currently I see a few alternatives to this:
1. One could call `calculate_MVWp_pole()` immediately after `calculate_pole_masses()`. In this way the value from the fit formula would be overwritten by the pure 1-loop calculation.
2. We add a flag to select which W mass to use in `calculate_pole_masses()`. This flag could be a parameter to `calculate_pole_masses()`, like so:

        void Standard_model::calculate_pole_masses(bool calculate_W_mass_from_fit = true) { ... }
3. Alternatively we could add a boolean flag as a member variable to `Standard_model`, which is used to select the calculation of the W mass in `calculate_pole_masses()`. (The default value should probably be `true` to not break the code in other places.)

It seems to me that the approach described in 3 is cleanest. What do you think? Maybe I have overlooked another possibility?

wkotlarski commented:
I'm not sure how this would work. The correct $m_W$ comes from matching routines. So it's disconnected from the pole mass calculation. I think that's why we used this hack. Because it's not easy to transfer it. Recomputation of $m_W$ from fit formula inside of pole masses only works for the SM because we only need the fit formula, nothing more. It wouldn't work for any other model.

wkotlarski commented:
I'm not sure how this would work. The correct $m_W$ comes from matching routines. So it's disconnected from the pole mass calculation. I think that's why we used this hack. Because it's not easy to transfer it. Recomputation of $m_W$ from fit formula inside of pole masses only works for the SM because we only need the fit formula. It wouldn't work for any other model.

wkotlarski commented:
Hi @Expander. Sorry, I've messed up. I forgot that there $m_W$ is computed correctly only if two scale matching is used. I've fixed the test and now the largest difference is 0.04%. I think we're good to go. I'd be grateful if you could have a look through this PR and decide if we can merge it. We can still tweak some small things regarding static linking etc on the development before the release.

Expander commented:
I'll be in the train to Aachen in 30 min and will have a look then.

Regarding the MW calculation: Which commit fixed the issue with the value of MW?

wkotlarski commented:
Thanks. It doesn't have to be today. $m_W$ was fixed on aac4306d05e7629c2f8bcb9da9d785464cd52b20. So long as the model is setup using two-scale matching everything is fine. So I've adopted the test to use two-scale matching instead of `setup_SM_const`. The biggest deviation right now is 0.04% and it probably comes mostly from numerical integration. The next one is 0.02% in $h \to Z \gamma$
```
   BOOST_CHECK_CLOSE(effc[0].gg.second, 1, 0.005);
   BOOST_CHECK_CLOSE(effc[0].gamgam.second, 1, 0.01);
   BOOST_CHECK_CLOSE(effc[0].Zgam.second, 1, 0.02);
   BOOST_CHECK_CLOSE(effc[0].ZZ.second, 1, 0.04);
   BOOST_CHECK_CLOSE(effc[0].WW.second, 1, 0.007);
   BOOST_CHECK_CLOSE(std::real(effc[0].ee.second),     1, 0.002);
   BOOST_CHECK_CLOSE(std::real(effc[0].mumu.second),   1, 0.002);
   BOOST_CHECK_CLOSE(std::real(effc[0].tautau.second), 1, 0.002);
   BOOST_CHECK_CLOSE(std::real(effc[0].bb.second), 1, 0.009);
   BOOST_CHECK_CLOSE(std::real(effc[0].cc.second), 1, 0.0010);
   BOOST_CHECK_CLOSE(std::real(effc[0].ss.second), 1, 0.0020);
   BOOST_CHECK_CLOSE(std::real(effc[0].dd.second), 1, 0.002);
   BOOST_CHECK_CLOSE(std::real(effc[0].uu.second), 1, 0.002);
```
To summarize: everything looks fine.

Expander commented on src/decays/specializations/H/decay_H_to_lbarl.inc:
```diff
@@ -41,5 +41,22 @@ double CLASSNAME::get_partial_width<Higgs, bar<ChargedLepton>::type, ChargedLept
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
+       static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_lilith)) ||
+       static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::calculate_normalized_effc))
```
Would you mind adapting these if-statements as
~~~.cpp
if (flexibledecay_settings.get(FlexibleDecay_settings::calculate_normalized_effc) != 0)
~~~

Expander commented on src/decays/specializations/H/decay_H_to_gg.inc:
```diff
@@ -165,6 +172,17 @@ double CLASSNAME::get_partial_width<Higgs, Gluon, Gluon>(
       );
    }
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
Same here.

Expander commented on meta/FSMathLink.m:
```diff
@@ -303,6 +369,39 @@
 
+PutEffCTableEntries[modelName_] :=
+    Module[{body = "", particleList = {}},
+       particleList = {{-1, 1, "dd"}, {-2, 2, "uu"}, {-3, 3, "ss"}, {-4, 4, "cc"}, {-5, 5, "bb"}, {-6, 6, "tt"}, {-11, 11, "ee"}, {-13, 13, "mumu"}, {-15, 15, "tautau"}, {-24, 24, "WW"}, {23, 23, "ZZ"}, {22, 23, "Zgam"}, {21, 21, "gg"}, {21, 22, "gamgam"}};
```
My proposal to create a local array of `std::pair<int,int>` could be used here: The `particleList` could be removed and one could access the struct of pairs. But then it may be best to put the array of pairs where it is globally accessible, maybe in `src/`.

Expander commented on meta/FSMathLink.m:
```diff
@@ -349,6 +448,26 @@
+           prototype = "void put_eff_couplings(MLINK link) const;\n";
+
+           body = "check_spectrum_pointer();\n" <>
+                  "auto& effc = get_normalized_higgs_effc();\n\n" <>
```
Can this local reference be made `const`?

wkotlarski replied:
Done in 144bb81de15d96fdcd6236f50be31aadc0ae1bf2.

Expander commented on meta/FSMathLink.m:
```diff
@@ -393,14 +512,16 @@
-            !static_cast<bool>(setting.get(flexiblesusy::Spectrum_generator_settings::calculate_bsm_masses))) {
+         auto settings = data.get_settings();
+         auto fdSettings = data.get_fd_settings();
+         if (!static_cast<bool>(settings.get(flexiblesusy::Spectrum_generator_settings::calculate_sm_masses)) ||
```
See my proposal to avoid the `static_cast<bool>` and write
~~~.cpp
if (settings.get(...) != 0) { ... }
~~~

Expander commented on README.rst:
I think it would be good to add a very small section with a proposal for the installation instructions of HiggsTools. I'm proposing this because one has to `install` HiggsTools. I did the following that worked:
~~~.sh
git clone https://gitlab.com/higgsbounds/higgstools.git
cd higgstools
mkdir build && cd build
HT=$PWD # HiggsTools build directory
cmake -DCMAKE_INSTALL_PREFIX=$HT/install ..
make
make install

cd $FS_DIR # FlexibleSUSY root directory
./configure --with-models=SM --with-higgstools-incdir=$HT/install/include --with-higgstools-libdir=$HT/src

~~~

wkotlarski replied:
Maybe 9d8812320bd934c72b26dcff365dc1c681b9a1c3 does the trick. There is nothing special that the user has to do when installing `HiggsTools` so I've decided to not include any details of its installation. This is the same what we did with `Collier` and `LoopTools`.

Expander replied:
I'm ok with that. 

One question, though: On my machine the path that contains the HiggsTools library is `lib/`, not `lib64/`. Should we add a short comment that the library path may be different?

Expander commented on src/decays/specializations/Ah/decay_Ah_to_AZ.inc:
```diff
@@ -23,6 +23,7 @@ double CLASSNAME::get_partial_width<PseudoscalarHiggs, Photon, ZBoson>(
       res *= alpha_em_0/alpha_em;
    }
 
+   std::string tag = field_as_string<Higgs>(in_idx) + "-" + field_as_string<Photon>(out1_idx) + "-" + field_as_string<ZBoson>(out2_idx);
```
Can this `tag` be made `const`?

Expander commented on src/decays/specializations/Ah/decay_Ah_to_AZ.inc:
```diff
@@ -32,5 +33,16 @@ double CLASSNAME::get_partial_width<PseudoscalarHiggs, Photon, ZBoson>(
       );
    }
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
See my comment on `static_cast<bool>`

wkotlarski replied:
Done in c587517522b7b65208aa28eb3ac3b70a03a3ce92.

Expander replied:
Thanks a lot!

Expander commented on src/decays/specializations/Ah/decay_Ah_to_AZ.inc:
```diff
@@ -23,6 +23,7 @@ double CLASSNAME::get_partial_width<PseudoscalarHiggs, Photon, ZBoson>(
+   std::string tag = field_as_string<Higgs>(in_idx) + "-" + field_as_string<Photon>(out1_idx) + "-" + field_as_string<ZBoson>(out2_idx);
    if (flexibledecay_settings.get(FlexibleDecay_settings::print_effc_block)) {
       effhiggscouplings_block_input.push_back(
          {
```
Could you please add a comment here, where these magic numbers (36,10,22,23) come from? One could alternatively move them into local variables with appropriate names.

wkotlarski replied:
I've replace this weird formula on 0b7454b155455c06b70a0088324e778152f08521. Maybe now it's more readable?

Expander replied:
This is much better! Now even I understand what this was about. :)

Expander commented on src/decays/specializations/Ah/decay_Ah_to_dbard.inc:
```diff
@@ -132,13 +132,29 @@ double CLASSNAME::get_partial_width<PseudoscalarHiggs, bar<DownTypeQuark>::type,
 
    const double result = (1-4.*xOS)*result_DR + 4*xOS*result_OS;
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
See my comment on `static_cast<bool>`

wkotlarski replied:
Done in c587517522b7b65208aa28eb3ac3b70a03a3ce92.

Expander commented on src/decays/specializations/Ah/decay_Ah_to_gg.inc:
```diff
@@ -87,5 +95,16 @@ double CLASSNAME::get_partial_width<PseudoscalarHiggs, Gluon, Gluon>(
       );
    }
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
See my comment on `static_cast<bool>`

wkotlarski replied:
Done in c587517522b7b65208aa28eb3ac3b70a03a3ce92.

Expander commented on src/decays/specializations/Ah/decay_Ah_to_lbarl.inc:
```diff
@@ -33,5 +33,22 @@ double CLASSNAME::get_partial_width<PseudoscalarHiggs, bar<ChargedLepton>::type,
       res *= 1. + get_alpha(context)/Pi*17./4.;
    }
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
See my comment on `static_cast<bool>`

wkotlarski replied:
Done in c587517522b7b65208aa28eb3ac3b70a03a3ce92.

Expander commented on src/decays/specializations/Ah/decay_Ah_to_ubaru.inc:
```diff
@@ -143,5 +143,21 @@ double CLASSNAME::get_partial_width<PseudoscalarHiggs, bar<UpTypeQuark>::type, U
       result = (1-4.*xOS)*result_DR + 4*xOS*result_OS;
    }
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
See my comment on `static_cast<bool>`

wkotlarski replied:
Done in c587517522b7b65208aa28eb3ac3b70a03a3ce92.

Expander commented on src/decays/specializations/Ah/decay_Ah_to_ubaru.inc:
```diff
@@ -143,5 +143,21 @@ double CLASSNAME::get_partial_width<PseudoscalarHiggs, bar<UpTypeQuark>::type, U
+         },
+         std::pair<std::string, std::complex<double>> {
+            field_as_string<PseudoscalarHiggs>(indexIn) + "-" + field_as_string<bar<UpTypeQuark>::type>(indexOut1) + "-" + field_as_string<UpTypeQuark>(indexOut2),
+            4.*std::max(xDR, xOS) > 1. ? std::imag(AhQQbarVertexDR_P)*1i : 1i*std::sqrt(result)
```
Could you please add a short comment here to explain this expression?

wkotlarski replied:
Done in 0659a0388bc16c58a1e9fb0110016ede73be7233.

Expander commented on src/decays/specializations/H/decay_H_to_AA.inc:
```diff
@@ -29,5 +29,16 @@ double CLASSNAME::get_partial_width<Higgs, Photon, Photon>(
       );
    }
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
See my comment on `static_cast<bool>`

wkotlarski replied:
Done in c587517522b7b65208aa28eb3ac3b70a03a3ce92.

Expander commented on src/decays/specializations/H/decay_H_to_AZ.inc:
```diff
@@ -19,15 +19,27 @@ double CLASSNAME::get_partial_width<Higgs, Photon, ZBoson>(
       res *= alpha_em_0/alpha_em;
    }
 
+   std::string tag = field_as_string<Higgs>(in_idx) + "-" + field_as_string<Photon>(out1_idx) + "-" + field_as_string<ZBoson>(out2_idx);
```
Can this `tag` be made `const`?

wkotlarski replied:
Done in 19f276db24c013e1a8b0c92bb00ede1d2418e08c.

Expander replied:
Thanks!

Expander commented on src/decays/specializations/H/decay_H_to_AZ.inc:
```diff
@@ -19,15 +19,27 @@ double CLASSNAME::get_partial_width<Higgs, Photon, ZBoson>(
    if (flexibledecay_settings.get(FlexibleDecay_settings::print_effc_block)) {
       effhiggscouplings_block_input.push_back(
          {
             25 + 10*(in_idx.size()>0 ? in_idx.at(0) : 0), 22, 23,
```
See my comment from above: Please add a brief comment to explain this expression.

Expander replied:
A reference would also be ok.

Expander replied:
I have the feeling that my comment refers to the wrong line. It applies to the line
~~~.cpp
25 + 10*...
~~~

wkotlarski replied:
This statement is no longer there.

Expander commented on src/decays/specializations/H/decay_H_to_AZ.inc:
```diff
@@ -19,15 +19,27 @@ double CLASSNAME::get_partial_width<Higgs, Photon, ZBoson>(
       );
    }
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
See my comment on `static_cast<bool>`

wkotlarski replied:
Done in c587517522b7b65208aa28eb3ac3b70a03a3ce92.

Expander commented on src/decays/specializations/H/decay_H_to_WpWm.inc:
```diff
@@ -95,5 +95,16 @@ double CLASSNAME::get_partial_width<Higgs, WpBoson, WmBoson>(
       res = flux * ps * mat_elem_sq;
    }
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
See my comment on `static_cast<bool>`

wkotlarski replied:
Done in c587517522b7b65208aa28eb3ac3b70a03a3ce92.

Expander commented on src/decays/specializations/H/decay_H_to_ZZ.inc:
```diff
@@ -102,5 +102,16 @@ double CLASSNAME::get_partial_width<Higgs,ZBoson,ZBoson>(
       res = flux * ps * ps_symmetry * mat_elem_sq;
    }
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
See my comment on `static_cast<bool>`

wkotlarski replied:
Done in c587517522b7b65208aa28eb3ac3b70a03a3ce92.

Expander commented on src/decays/specializations/H/decay_H_to_ubaru.inc:
```diff
@@ -177,7 +177,24 @@ double CLASSNAME::get_partial_width<Higgs, bar<UpTypeQuark>::type, UpTypeQuark>(
-}
+   }
+
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
See my comment on `static_cast<bool>`

Expander commented on test/SOFTSUSY/module.mk:
```diff
@@ -115,11 +115,11 @@ $(LIBSOFTSUSY): $(LIBSOFTSUSY_OBJ)
 $(RUN_SOFTSUSY_EXE): $(RUN_SOFTSUSY_OBJ) $(LIBSOFTSUSY) $(MODSOFTSUSY_LIB) $(LIBFLEXI) $(filter-out -%,$(LOOPFUNCLIBS)) $(FUTILIBS)
 		@$(MSG)
-		$(Q)$(CXX) -o $@ $(call abspathx,$^) $(filter -%,$(LOOPFUNCLIBS)) $(FLIBS)
+		$(Q)$(CXX) -o $@ $(call abspathx,$^) $(filter -%,$(LOOPFUNCLIBS)) $(FLIBS) $(PYTHONLIBS) ${GSLLIBS} ${HIGGSTOOLSLIBS}
```
I think we should use parenthesis here, i.e.
~~~
$(GSLLIBS) $(HIGGSTOOLSLIBS)
~~~

wkotlarski replied:
Fixed in debf2db59b3ce41e7fff09498d9ec68a7c5669f5.

Expander commented on test/SOFTSUSY/module.mk:
```diff
@@ -115,11 +115,11 @@ $(LIBSOFTSUSY): $(LIBSOFTSUSY_OBJ)
 $(RUN_SOFTPOINT_EXE): $(RUN_SOFTPOINT_OBJ) $(LIBSOFTSUSY) $(MODSOFTSUSY_LIB) $(LIBFLEXI) $(filter-out -%,$(LOOPFUNCLIBS)) $(FUTILIBS)
 		@$(MSG)
-		$(Q)$(CXX) -o $@ $(call abspathx,$^) $(filter -%,$(LOOPFUNCLIBS)) $(FLIBS)
+		$(Q)$(CXX) -o $@ $(call abspathx,$^) $(filter -%,$(LOOPFUNCLIBS)) $(FLIBS) $(PYTHONLIBS) ${GSLLIBS} ${HIGGSTOOLSLIBS}
```
I think we should use parenthesis here, i.e.
~~~
$(GSLLIBS) $(HIGGSTOOLSLIBS)
~~~

wkotlarski replied:
Fixed in debf2db59b3ce41e7fff09498d9ec68a7c5669f5.

Expander commented on src/decays/experimental_constraints.cpp:
```diff
@@ -0,0 +1,579 @@
+}
+
+void print_effc(double mass, HP::NeutralEffectiveCouplings const& effC) {
+   std::cout << "Effective couplings for particle of mass " << mass << '\n';
```
Please do not end every `cout` statement with `std::endl`, because this is very slow. Please use `'\n'` instead. It would be ok to `std::endl` in the very last line, though.

Reason: `std::endl` calls `coud.flush()`, which is quite expensive and which is not needed at the end of every line here.

wkotlarski replied:
This is not needed. This was just some dubug helper function. I've removed it in 9166aaaae3b6d65ce5464f5c991a0ec16078a238.

Expander commented on src/decays/experimental_constraints.cpp:
```diff
@@ -0,0 +1,579 @@
+   // Creating an object of the class Lilith: lilithcalc
+   PyObject* lilithcalc2 = initialize_lilith(experimental_input);
+
+   char XMLinputstring[6000] = "";
```
Sorry to be pedantic, but I'm afraid of filling these fixed-size buffers. Please use a `std::string` or `std::sstream` instead, which is dynamically re-sized. We can always use `.c_str()` to get a C-style character array out, if we need it at the end.

wkotlarski replied:
Done in 4f9b9025de97dd9aaac54036b32101e380fe7fc7.

Expander replied:
Excellent, thanks!

Expander commented on src/decays/experimental_constraints.cpp:
```diff
@@ -0,0 +1,579 @@
+
+      int status, iter = 0;
+      static constexpr int max_iter = 100;
+      const gsl_min_fminimizer_type *T;
```
Could you please open another issue to refactor this C-style minimization using GSL? This workaround and the manual allocation and (not exception-safe) manual deallocation makes me nervous. :)

wkotlarski replied:
Done (#641)

Expander commented on src/decays/experimental_constraints.cpp:
```diff
@@ -0,0 +1,579 @@
+   // built in SM
+   auto flexibledecay_settings_ = flexibledecay_settings;
+
+   EffectiveCoupling_list _bsm_input;
```
We usually don't use variable names that begin with an underscore `_`. In particular local variables can have regular names. (There is an exception for constructor arguments, though.)

wkotlarski replied:
Done in 167a4c3e1e4876f10692ce55b054c26b4fa921f3.

Expander commented on src/decays/experimental_constraints.cpp:
```diff
@@ -0,0 +1,579 @@
+         // fermion channels are given as complex numbers
+         // we normalize to real part of SM coupling
+         // quarks
+         NeutralHiggsEffectiveCouplings _coups {el};
```
Same comment here: Please do not use an underscore as first character for this local variable.

wkotlarski replied:
Done in 167a4c3e1e4876f10692ce55b054c26b4fa921f3.

Expander commented on src/decays/experimental_constraints.cpp:
```diff
@@ -0,0 +1,579 @@
+      // for mh > 700 GeV this gives λ > 4
+      // it probably makes no sense to use coupling strengh modifiers in this case so we skip those particles
+      // On the other hand there's a problem with finding a SM equivalent of very light states
+      if (mass > 650 || mass < 1) continue;
```
I'd recommend to also print a warning here. We could use `WARNING` or `VERBOSE_MSG`.

wkotlarski replied:
Done in bae5f3cea1ae2364a821757bf2a81c158a9f60e3. I decided to go with `VERBOSE_MSG` because it will happen in almost all models so it doesn't make sense to pollute the output. I've added the following message
```
Cannot construct normalization for effective coupling of hh(2) with mass of 19856.713128 GeV
```

Expander replied:
Thanks!

Expander commented on src/decays/experimental_constraints.cpp:
```diff
@@ -0,0 +1,579 @@
+   }
+
+   const auto signals = Higgs::Signals {higgssignals_dataset};
+   //for (const auto &m : signals.measurements()) {
```
Please remove this debug output or use `VERBOSE_MSG`.

wkotlarski replied:
Done in 537b68cbcc9fb7bee3e3a6a53ef154f2123eca05.

Expander commented on src/decays/experimental_constraints.cpp:
```diff
@@ -0,0 +1,579 @@
+   for (auto const& el : bsm_input) {
+      const double mh = el.mass;
+      if (mh > 123.0 && mh < 128.0) {
+         higgs_in_range = true || higgs_in_range;
```
This statements can be shortened to
~~~.cpp
higgs_in_range = true;
~~~

wkotlarski replied:
Done in ad89f6dc3a33ee105a10ac5f8f402ae96272d408.

Expander commented on src/decays/experimental_constraints.cpp:
```diff
@@ -0,0 +1,579 @@
+   // Creating an object of the class Lilith: lilithcalc
+   PyObject* lilithcalc = initialize_lilith(const_cast<char*>(lilith_db.c_str()));
+
+   char XMLinputstring[6000]="";
```
See my comment on the constant character buffer issue: Please use a `std::string` or `std::sstream` for filling.

wkotlarski replied:
Done in abc15f452820999c33bcad3254a732ef13e5b4b6.

wkotlarski commented on src/decays/specializations/Ah/decay_Ah_to_AA.inc:
```diff
@@ -33,5 +33,16 @@ double CLASSNAME::get_partial_width<PseudoscalarHiggs, Photon, Photon>(
       );
    }
 
+   if (static_cast<bool>(flexibledecay_settings.get(FlexibleDecay_settings::call_higgstools)) ||
```
Done in c587517522b7b65208aa28eb3ac3b70a03a3ce92.

wkotlarski commented on src/decays/experimental_constraints.hpp:
```diff
@@ -0,0 +1,80 @@
+ * HiggsTools and Lilith
+ */
+
+#ifndef HIGGSTOOLS_INTERFACE_H
```
Done in c10d8d6aa8ef3abc7da58ffa367d19bd76460f37.

Expander commented:
Sorry, I can't find the struct. Where is it defined?

wkotlarski commented:
I was confused because this doesn't work with strongly typed enums. We'd have to do an explicit cast. Can you please check if b58052c37ab39d520feb66b239f6a49a52296bf3 does what you wanted?

Expander commented:
Thanks, that looks perfect!

One small thing regarding the naming, though: I'd say that a PDG ID is a single number (not a pair of numbers). Currently the enum and the array are called `PDG_id` and `pdg_id`, respectively. I think a name that reflects the fact that they refer to pairs of ids would be better. What about something like `PDG_id_pairs` and `pdg_id_pairs` or similar?

Expander commented on src/decays/decay.cpp:
```diff
@@ -133,4 +155,174 @@ double hVV_4body(double *q2, size_t /* dim */, void *params)
+         effC.mumu = c;
+      }
+      else if (are_the_same(fs, pdg_id_pairs[PDG_id_pairs::tautau])) {
+         std::cout << fs[0] << ' ' << fs[1] << std::endl;
```
It seems that this is a left-over comment from debugging. I guess it should be removed.

Expander commented on src/decays/decay.cpp:
```diff
@@ -46,6 +46,28 @@ std::size_t hash_pid_list(int pid_in, Container pids_out)
 
 } // anonymous namespace
 
+extern int PDG_id_pairs[PDG_id_pairs::NUMBER_OF_PDG_IDS];
```
I think this line is incorrect and must be removed. (We define an enum class in the header with the same name.)

wkotlarski commented:
@Expander, do you think that we could now merge this PR?
